src/charm.py

```diff
@@ -612,26 +730,26 @@ def _get_configs_per_leader(self, relation: Relation) -> Optional[dict]:
             data = ipa.get_data(relation)
         except DataValidationError as e:
             logger.error(f"invalid data shared through {relation}... Error: {e}.")
-            return None
+            return {}
 
-        prefix = self._get_prefix(data)
-        config = self._generate_per_leader_config(prefix, data)
+        prefix = self._get_prefix(data)  # type: ignore
+        config = self._generate_per_leader_config(prefix, data)  # type: ignore
         if self.unit.is_leader():
             ipa.publish_url(relation, self._get_external_url(prefix))
 
         return config
 
-    def _get_configs_per_app(self, relation: Relation) -> Optional[dict]:
+    def _get_configs_per_app(self, relation: Relation) -> Dict[str, Any]:
```

Comment:
Should we have a new recursive type e.g. StrDict?
```python
StrDict = Dict[str, Union[str, "StrDict"]]
```

Reply:
elsewhere I had an AnyJson type, probably a better fit as there can be booleans, arrays and integers as well

Reply:
all type checkers are fussy with recursive types either way in my experience. 
Gave it half an hour to try out adding an AnyJson type but it got way worse. Dropping it for now, if we like we can have a separate pr to improve the typing of traefik as a whole, but that'd feel like feature creep in this context.